### PR TITLE
REST APIs for Gobblin Service job configuration.

### PR DESCRIPTION
**Pull request by htran1**
Create APIs for create, get, full update, and delete of JobConfigs.
JobConfigs are turned into JobSpecs stored in a MutableJobCatalog.

### CODE REVIEW

**Comment by htran1** [comment]
@ibuenros @abti, please take a look. I pulled the files with changes requested in PR #1571 into a new branch since the other PR was polluted with changes from other checkins.

File: gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-server/src/main/java/gobblin/service/FlowConfigsResource.java [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+import gobblin.runtime.api.MutableJobCatalog;
+import gobblin.runtime.job_catalog.FSJobCatalog;
+import gobblin.runtime.job_catalog.ImmutableFSJobCatalog;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Properties;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
Comment (ibuenros) [inline gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-server/src/main/java/gobblin/service/FlowConfigsResource.java]: Please follow import conventions.

File: gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-server/src/main/java/gobblin/service/FlowConfigsResource.java [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+import static org.reflections.util.ConfigurationBuilder.*;
```
Comment (ibuenros) [inline gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-server/src/main/java/gobblin/service/FlowConfigsResource.java]: No * imports

File: gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/idl/gobblin.service.flowconfigs.restspec.json [hunk 3]
```diff
@@ -0,0 +1,31 @@
+{
+  "name" : "flowconfigs",
+  "namespace" : "gobblin.service",
+  "path" : "/flowconfigs",
+  "schema" : "gobblin.service.FlowConfig",
+  "doc" : "generated from: gobblin.service.FlowConfigsResource",
```
Comment (li-ahamstra) [inline gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/idl/gobblin.service.flowconfigs.restspec.json]: Please add a javadoc to the resource. That should automatically create a doc here.

File: gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/idl/gobblin.service.flowconfigs.restspec.json [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  "collection" : {
+    "identifier" : {
+      "name" : "flowconfigsId",
```
Comment (li-ahamstra) [inline gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/idl/gobblin.service.flowconfigs.restspec.json]: Could just be id, but definitely change to flowconfigId

File: gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/idl/gobblin.service.flowconfigs.restspec.json [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      "type" : "gobblin.service.FlowConfigId",
+      "params" : "com.linkedin.restli.common.EmptyRecord"
+    },
+    "supports" : [ "create", "delete", "get", "update" ],
+    "methods" : [ {
+      "method" : "create",
+      "doc" : "Put a new {@link JobSpec} based on flowConfig in the {@link MutableJobCatalog}"
```
Comment (li-ahamstra) [inline gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/idl/gobblin.service.flowconfigs.restspec.json]: Document from the perspective of the API user.

File: gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/idl/gobblin.service.flowconfigs.restspec.json [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    }, {
+      "method" : "get",
+      "doc" : "Retrieve the {@link FlowConfig} with the given key"
+    }, {
+      "method" : "update",
+      "doc" : "Update the {@link JobSpec} in the {@link MutableJobCatalog} based on the specified {@link FlowConfig}"
```
Comment (li-ahamstra) [inline gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/idl/gobblin.service.flowconfigs.restspec.json]: Same comment as above...document for the user. Is upsert an option?

Comment (li-ahamstra) [inline gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/idl/gobblin.service.flowconfigs.restspec.json]: Also, document which of the jobs that result from this config are affected by the changes.

File: gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/idl/gobblin.service.flowconfigs.restspec.json [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    }, {
+      "method" : "delete",
+      "doc" : "delete a configured flow"
```
Comment (li-ahamstra) [inline gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/idl/gobblin.service.flowconfigs.restspec.json]: What happens to running and scheduled jobs?

File: gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/pegasus/gobblin/service/FlowConfig.pdsc [hunk 8]
```diff
@@ -0,0 +1,38 @@
+{
+  "type" : "record",
+  "name" : "FlowConfig",
+  "namespace": "gobblin.service",
+  "doc" : "Configuration for a flow",
```
Comment (li-ahamstra) [inline gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/pegasus/gobblin/service/FlowConfig.pdsc]: Tell me something that isn't already in the name.

File: gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/pegasus/gobblin/service/FlowConfig.pdsc [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    },
+    {
+      "name" : "flowGroup",
+      "type" : "string",
+      "doc" : "Group of the flow. This defines the namespace for the flow."
+    },
+    {
+      "name" : "schedule",
+      "type" : "string",
+      "doc" : "Schedule for flow in cron format"
```
Comment (li-ahamstra) [inline gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/pegasus/gobblin/service/FlowConfig.pdsc]: Think about choosing a typeref for validation or using a validator directly here.

File: gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/pegasus/gobblin/service/FlowConfig.pdsc [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    },
+    {
+      "name" : "runImmediately",
+      "type" : "boolean",
+      "optional" : true,
```
Comment (li-ahamstra) [inline gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/pegasus/gobblin/service/FlowConfig.pdsc]: default false instead

File: gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/pegasus/gobblin/service/FlowConfig.pdsc [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+      "doc" : "Set to true to request that job be run immediately"
```
Comment (li-ahamstra) [inline gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/pegasus/gobblin/service/FlowConfig.pdsc]: ...as well as following the schedule.

File: gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/pegasus/gobblin/service/FlowConfig.pdsc [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+      "doc" : "Name of the flow"
+    },
+    {
+      "name" : "flowGroup",
+      "type" : "string",
+      "doc" : "Group of the flow. This defines the namespace for the flow."
+    },
+    {
+      "name" : "schedule",
+      "type" : "string",
```
Comment (li-ahamstra) [inline gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/pegasus/gobblin/service/FlowConfig.pdsc]: Set a default or optional...means run once. Maybe a union? enum? Make the schedule an evolvable type.

File: gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/pegasus/gobblin/service/FlowConfig.pdsc [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    {
+      "name" : "runImmediately",
+      "type" : "boolean",
+      "optional" : true,
+      "doc" : "Set to true to request that job be run immediately"
+    },
+    {
+      "name" : "templateUris",
+      "type" : "string",
+      "doc" : "Comma separated list of URIs for templates used in the flow"
```
Comment (li-ahamstra) [inline gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/pegasus/gobblin/service/FlowConfig.pdsc]: Document how a user can figure out what the URIs are. Think about a URI typeref for validation.

File: gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/pegasus/gobblin/service/FlowConfig.pdsc [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    },
+    {
+      "name" : "properties",
+      "type" : { "type" : "map", "values" : "string" },
+      "doc" : "Properties for the flow"
```
Comment (li-ahamstra) [inline gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/pegasus/gobblin/service/FlowConfig.pdsc]: Ensure that the properties that are encompassed by this field are only properties that are not understood by GaaS.

File: gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/pegasus/gobblin/service/FlowConfigId.pdsc [hunk 15]
```diff
@@ -0,0 +1,17 @@
+{
+  "type" : "record",
+  "name" : "FlowConfigId",
+  "namespace": "gobblin.service",
+  "doc" : "Identifier for a flow",
+  "fields" : [
+    {
+      "name" : "flowName",
+      "type" : "string",
+      "doc" : "Name of the flow"
```
Comment (li-ahamstra) [inline gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-api/src/main/pegasus/gobblin/service/FlowConfigId.pdsc]: May want validation, especially for length.

**Pull request merged**